File: solution/90/solution.py
import copy
class Solution(object):
    def subsetsWithDup(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        nums.sort()
        item = []
        result = [[]]
        self.generate(0, nums, item, result)
        return result

    def generate(self, i, nums, item, result):
        if i >= len(nums):return
        item.append(nums[i])
        if item not in result:result.append(copy.copy(item))
        self.generate(i+1, nums, item, result)
        item.pop()
        self.generate(i+1, nums, item, result)


s = Solution()
print s.subsetsWithDup([1,2,2])

File: solution/90/solution.cpp
class Solution {
public:

    void generate(int i, vector<int>& nums, set<vector<int>>& res_set, vector<int>& item, vector<vector<int>>& result){
        if(i>=nums.size())return;
        item.push_back(nums[i]);
        if(res_set.find(item)==res_set.end()){result.push_back(item);res_set.insert(item);}
        generate(i+1, nums, res_set, item, result);
        item.pop_back();
        generate(i+1, nums, res_set, item, result);
    }

    vector<vector<int>> subsetsWithDup(vector<int>& nums) {
        vector<int> item;
        vector<vector<int>> result;
        set<vector<int>> res_set;
        result.push_back(item);
        std::sort(nums.begin(),nums.end());
        generate(0, nums, res_set, item, result);
        return result;
    }
};